src/data/eventsource.ts

```diff
@@ -0,0 +1,255 @@
+ */
+export class ConnectionFailedError extends Error {
+  readonly name = 'ConnectionFailedError'
+}
+
+/**
+ * The listener has been told to explicitly disconnect.
+ *  This is a rare situation, but may occur if the API knows reconnect attempts will fail,
+ *  eg in the case of a deleted dataset, a blocked project or similar events.
+ * @public
+ */
+export class DisconnectError extends Error {
+  readonly name = 'DisconnectError'
+  readonly reason?: string
+  constructor(message: string, reason?: string, options: ErrorOptions = {}) {
+    super(message, options)
+    this.reason = reason
+  }
+}
+
+/**
+ * @public
+ * The server sent a `channelError` message. Usually indicative of a bad or malformed request
+ */
+export class ChannelError extends Error {
+  readonly name = 'ChannelError'
+  readonly data?: unknown
+  constructor(message: string, data: unknown) {
+    super(message)
+    this.data = data
+  }
+}
+
+/**
+ * @public
+ * The server sent an `error`-event to tell the client that an unexpected error has happened.
+ */
+export class MessageError extends Error {
+  readonly name = 'MessageError'
+  readonly data?: unknown
+  constructor(message: string, data: unknown, options: ErrorOptions = {}) {
+    super(message, options)
+    this.data = data
+  }
+}
+
+/**
+ * @public
+ * An error occurred while parsing the message sent by the server as JSON. Should normally not happen.
+ */
+export class MessageParseError extends Error {
+  readonly name = 'MessageParseError'
+}
+
+/**
+ * @public
+ */
+export interface ServerSentEvent<Name extends string> {
+  type: Name
+  id?: string
+  data?: unknown
+}
+
+// Always listen for these events, no matter what
+const REQUIRED_EVENTS = ['channelError', 'disconnect']
+
+/**
+ * @internal
+ */
+export type EventSourceEvent<Name extends string> = ServerSentEvent<Name>
+
+/**
+ * @internal
+ */
+export type EventSourceInstance = InstanceType<typeof globalThis.EventSource>
+
+/**
+ * Sanity API specific EventSource handler shared between the listen and live APIs
+ *
+ * Since the `EventSource` API is not provided by all environments, this function enables custom initialization of the EventSource instance
+ * for runtimes that requires polyfilling or custom setup logic (e.g. custom HTTP headers)
+ * via the passed `initEventSource` function which must return an EventSource instance.
+ *
+ * Possible errors to be thrown on the returned observable are:
+ * - {@link MessageError}
+ * - {@link MessageParseError}
+ * - {@link ChannelError}
+ * - {@link DisconnectError}
+ * - {@link ConnectionFailedError}
+ *
+ * @param initEventSource - A function that returns an EventSource instance or an Observable that resolves to an EventSource instance
+ * @param events - an array of named events from the API to listen for.
+ *
+ * @internal
+ */
+export function connectEventSource<EventName extends string>(
```

Comment:
nit:
```suggestion
const EventTypeName = ListenEvent['type']
export function connectEventSource<EventName extends EventTypeName[]>(
```

should make this type safe? 🤔  and same on L123?

Reply:
It is type safe… kind of. Here we can only really know for _sure_ that the event matches any of the event names given, which is why this returns `{type: Name, id:? string, data?: unknown}`. I left it up to the endpoints themselves to take care of typing it correctly, since they have more specific knowledge of the payloads. I also made a choice not to splat `data` into the events at this level, because in theory it could contain `type` and `id` which would lead to very surprising behavior.